Handle case where user sends GC as gift via PPDG

Enuratique opened the pull request:
When a user sends a gift card as a gift instead of buying for one's self, PPDG renders some additional content before rendering the skip button. As such, in its current state, the code is evaluated before there's a skip button to click. This amendment attaches a DOM observer and will attempt the skip button logic for each time a dom element is added or removed. Attribute changes are ignored. An initial attempt at clicking the skip button is still made for pages where this is not an issue.

Enuratique committed:
Handle case where user sends GC as gift via PPDG

When a user sends a gift card as a gift instead of buying for one's self, PPDG renders some additional content before rendering the skip button. As such, in its current state, the code is evaluated before there's a skip button to click. This amendment attaches a DOM observer and will attempt the skip button logic for each time a dom element is added or removed. Attribute changes are ignored. An initial attempt at clicking the skip button is still made for pages where this is not an issue.

Enuratique committed:
Fixed possible bug / leak with orphaned observer

Rearranged order of button clicking and observer detaching so that the observer is disconnected and killed before navigating away from that page

Enuratique committed:
If statement was missing curly braces

Lol this is why you always leave a note... Err, always use curly braces for your if statements

mbianucci commented on skip.js:
```diff
@@ -1,18 +1,70 @@
+
+    // Callback function to execute when mutations are observed
+    const callback = function(mutationsList, observer) {
+        // Use traditional 'for loops' for IE 11
```
Probably wouldn't need to support IE 11 from a Chromium/Firefox extension, but I was using the same MDN page as reference :)

Enuratique replied:
Yeah this was copy pasta from that MDN page... You can disregard that or even take away attributes from the config and remove that else if clause... I figure it doesn't hurt to have in case more sites go this route in the future - the plumbing is already there albeit commented out.

mbianucci commented on skip.js:
```diff
@@ -1,18 +1,70 @@
+function init() {
+    switch (window.location.hostname) {
+        case 'delivery.cardago.com':
+            configure_dom_observer('content-main');
```
Afaict, `content-main` doesn't mutate at all on MGCP GC. I couldn't get a mutation observer to fire when I added it by hand in the console, do you know if it normally would if it was part of the extension? Otherwise this might never click the button.

Enuratique replied:
So yes you're right, attaching an observer for MGCP is overkill since it doesn't mutate like PPDG. But I like creating one-size-fits-all since the logic just gets simpler, especially if more sites need to be supported in the future. The flow of the logic is in init, we explicitly call go() at the end after the observer has been instantiated anyways, so there's always a first attempt to click a matching button if one exists. For PPDG, this go() call will do nothing since the buttons array will be empty when we explicitly call it, and then later go() will be called for each mutation, which eventually will have the matching button in the DOM (and it may also be called a few more times if there are mutations prior to the Skip button appearing). 

On the flip side, MGCP we set up an observer which will immediately be disconnected since the matching button will be found in the first attempt. IMO the computational overhead of calling go() superfluously and/or setting up an observer that will never observe a mutation is a worthy tradeoff for simpler code logic.

mbianucci replied:
Agreed, I had missed the go() call at the end of this function when I had commented. I definitely strive to avoid having any site specific things as well, making things as general as possible is the goal, so I like what you were going for here.

The one thing that I'm not sure of is how to avoid the observer being global in that case, since we wouldn't have the callback getting the observer to pass to go. Although thinking about it now, seems like we could probably just create the observer in init() and pass it to configure_dom_observer() and go() as necessary and it wouldn't be a problem.

mbianucci commented on skip.js:
```diff
@@ -1,18 +1,70 @@
+        for(let mutation of mutationsList) {
+            if (mutation.type === 'childList') {
+                //console.log('A child node has been added or removed.');
+                go();
```
Could avoid the observer being a global by passing it along here, I think.

Enuratique replied:
Yeah globals are probably a bad idea, you're right. Just being lazy.

mbianucci commented on skip.js:
```diff
@@ -1,18 +1,70 @@
+    go();
 }
+
+init();
```
Is this the normal best practice for javascript, instead of a main() like I did? I just searching around a little last night for best practice stuff, but every where you look you see conflicting information, especially about something like best practices.

Enuratique replied:
Really there's no difference between main vs init... They both get called implicitly by being the last top-level scoped code during evaluation... You previously had the switch statement instead of putting it in a function and calling the function in the switch statement's place. I'm not sure if it falls under best practice, but like global variables, having code just hang out like that is also frowned upon when it can be wrapped in a function which provides at least a modicum of variable scoping.

mbianucci replied:
Gotcha, I didn't think there was a different, just wasn't sure if it was more common for JS to have an init() function instead of a main().

mbianucci commented:
I think you've got a lot of good ideas going on here - if you want to resolve the merge conflicts and start a review, I'd be more than happy to review it and get landed. If not, I can do it on my end, actually start a PR (instead of just pushing to master!), and you can take a look too. Either one is fine with me.

Enuratique committed:
Refactored global observer, changed observer config, cleaned up code

Removed superfluous attribute observer configuration

Changed tabs from 4 to 2 spaces

Removed global observer variable

mbianucci committed:
Merge branch 'master' into patch-1
